src/librustc_middle/ty/trait_def.rs

```diff
@@ -187,32 +187,38 @@ pub(super) fn all_local_trait_impls<'tcx>(
 pub(super) fn trait_impls_of_provider(tcx: TyCtxt<'_>, trait_id: DefId) -> TraitImpls {
     let mut impls = TraitImpls::default();
 
-    {
-        let mut add_impl = |impl_def_id: DefId| {
-            let impl_self_ty = tcx.type_of(impl_def_id);
```

Comment:
so is the performance improvement here just that we don't have to compute `type_of` and then simplify the type, but instead we do it "up front" when encoding?

Reply:
Right, I think all the cost is in decoding the `Ty` from `type_of` (simplification should be relatively cheap).

And based on the detailed query profiling info, I'm guessing it's `ty::Adt` containing a `&'tcx ty::AdtDef` instead of just a `DefId`, and somehow `item_attrs` getting called for every `struct`/`enum`/`union` implementing the trait.

I'm suspecting this is similar to what @nnethercote was talking about a few months ago, where the deserialization of doc comment attributes was a significant cost (I don't recall what happened with that line of investigation though).

Reply:
#65750 introduced a specialized representation for doc comments, that might be what you are referring to.